src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt

```diff
@@ -0,0 +1,78 @@
+package com.safehill.kcrypto
+
+import dev.turingcomplete.kotlinonetimepassword.HmacAlgorithm
+import dev.turingcomplete.kotlinonetimepassword.TimeBasedOneTimePasswordConfig
+import dev.turingcomplete.kotlinonetimepassword.TimeBasedOneTimePasswordGenerator
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.toJavaDuration
+
+/**
+ * @param digits number of digits in the otp.
+ * @param validDuration the validity of otp.
+ * The generated otp will be the same for half the time step if the same secret is used.
+ * Use [isValid] to check if the otp is valid within the validity duration.
+ */
+class SafehillOTP(
+    digits: Int,
+    validDuration: Duration
+) {
+    private val halfTimeStep = validDuration / 2
+
+    private val otpConfig = TimeBasedOneTimePasswordConfig(
+        codeDigits = digits,
+        hmacAlgorithm = HmacAlgorithm.SHA1,
+        timeStep = halfTimeStep.inWholeMicroseconds,
+        timeStepUnit = TimeUnit.MICROSECONDS
+    )
+
+    /**
+     * OTP code generation
+     * @param secret the cryptographically-secure secret
+     * @param digits how long (the number of digits) is the OTP
+     * @param timeStep the validity of otp
+     * @return [OTPInfo] holding the code and validity duration.
+     */
+    fun generateCode(
+        secret: ByteArray
+    ): OTPInfo {
+        val generator = TimeBasedOneTimePasswordGenerator(secret, otpConfig)
+        val otpCode = generator.generate()
+
+        val counter = generator.counter()
+        val endEpochMillis = generator.timeslotStart(counter + 2) - 1
+        return OTPInfo(
+            otp = otpCode,
+            validity = (endEpochMillis - Instant.now().toEpochMilli()).milliseconds
+        )
+    }
+
+
+    /**
+     * @param secret the secret that was used to generate the otp code.
+     * @param code the code to check whether it is valid or not
+     * @return true if the code is valid, false otherwise
+     */
+    fun isValid(
```

Comment:
I wouldn't bake this split logic here, but in the client of this API

Reply:
Do I add a slot parameter to generate the OTP?
O refering to the current time slot.
1 refering to the next slot. 
-1 refering to the previous slot. 

Should the logic we discussed yesterday (where we generate otp valid for x/2 duration and check current slot and previous slot) reside in the client side as well?

Reply:
Yes, and you can pass the instant